### PR TITLE
[SPARK] Extract DatasetIdentifier from SaveIntoDataSourceCommandVisitor options

### PR DESCRIPTION
**Pull request by ddebowczyk92**
### Problem

Closes: #2921 

### Solution

Extracts DatasetIdentifier from command's options instead of relying on `p.createRelation(sqlContext, command.options())` which  is a heavy operation for `JdbcRelationProvider`

> **Note:** All schema changes require discussion. Please [link the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) for context.

- [ ] Your change modifies the [core](https://github.com/OpenLineage/OpenLineage/blob/main/spec/OpenLineage.json) OpenLineage model
- [ ] Your change modifies one or more OpenLineage [facets](https://github.com/OpenLineage/OpenLineage/tree/main/spec/facets)

If you're contributing a new integration, please specify the scope of the integration and how/where it has been tested (e.g., Apache Spark integration supports `S3` and `GCS` filesystem operations, tested with AWS EMR).

### Checklist

- [x] You've [signed-off](https://github.com/OpenLineage/OpenLineage/blob/main/why-the-dco.md) your work
- [x] Your pull request title follows our [guidelines](https://github.com/OpenLineage/OpenLineage/blob/main/CONTRIBUTING.md#creating-pull-requests)
- [x] Your changes are accompanied by tests (_if relevant_)
- [x] Your change contains a [small diff](https://kurtisnusbaum.medium.com/stacked-diffs-keeping-phabricator-diffs-small-d9964f4dcfa6) and is self-contained
- [ ] You've updated any relevant documentation (_if relevant_)
- [ ] Your comment includes a one-liner for the changelog about the specific purpose of the change (_not required for changes to tests, docs, or CI config_)
- [ ] You've versioned the core OpenLineage model or facets according to [SchemaVer](https://docs.snowplowanalytics.com/docs/pipeline-components-and-applications/iglu/common-architecture/schemaver) (_if relevant_)
- [ ] You've added a [header](https://github.com/OpenLineage/OpenLineage/tree/main/.github/header_templates.md) to source files (_if relevant_)

----
SPDX-License-Identifier: Apache-2.0\
Copyright 2018-2024 contributors to the OpenLineage project

### CODE REVIEW

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/lifecycle/plan/SaveIntoDataSourceCommandVisitor.java [hunk 1]
```diff
@@ -143,6 +146,21 @@ public List<OutputDataset> apply(SparkListenerEvent event, SaveIntoDataSourceCom
       }
     }
 
+    if (command
+        .dataSource()
+        .getClass()
+        .getCanonicalName()
+        .equals(JdbcRelationProvider.class.getCanonicalName())) {
+      if (command.options().contains("dbtable") && command.options().contains("url")) {
```
Comment (Imbruced) [inline integration/spark/shared/src/main/java/io/openlineage/spark/agent/lifecycle/plan/SaveIntoDataSourceCommandVisitor.java:154]: is it possible that options are null ?

**Comment by Imbruced** [comment]
it looks good to me :) But I have no such power to accept MR :)

File: integration/spark/app/src/test/java/io/openlineage/spark/agent/lifecycle/plan/SaveIntoDataSourceCommandVisitorTest.java [hunk 2]
```diff
@@ -53,6 +58,16 @@ class SaveIntoDataSourceCommandVisitorTest {
...
   SparkOpenLineageExtensionVisitorWrapper wrapper =
       mock(SparkOpenLineageExtensionVisitorWrapper.class);
+
+  SparkListenerEvent event = mock(SparkListenerEvent.class);
+
+  StructType schema =
+      new StructType(
+          new StructField[] {
+            new StructField("key", IntegerType$.MODULE$, false, new Metadata(new HashMap<>())),
+            new StructField("value", StringType$.MODULE$, false, new Metadata(new HashMap<>()))
```
Comment (d-m-h) [inline integration/spark/app/src/test/java/io/openlineage/spark/agent/lifecycle/plan/SaveIntoDataSourceCommandVisitorTest.java:68]: Something slightly easier on the eyes is to use:

```suggestion
              new StructField("key", DataTypes.IntegerType, false, new Metadata(new HashMap<>())),
            new StructField("value", DataTypes.StringType, false, new Metadata(new HashMap<>()))
```

File: integration/spark/app/src/test/java/io/openlineage/spark/agent/lifecycle/plan/SaveIntoDataSourceCommandVisitorTest.java [hunk 3]
```diff
@@ -32,13 +32,18 @@
 import org.apache.spark.sql.catalyst.expressions.Attribute;
 import org.apache.spark.sql.catalyst.plans.logical.LocalRelation;
 import org.apache.spark.sql.execution.datasources.SaveIntoDataSourceCommand;
+import org.apache.spark.sql.execution.datasources.jdbc.JdbcRelationProvider;
 import org.apache.spark.sql.sources.CreatableRelationProvider;
 import org.apache.spark.sql.sources.RelationProvider;
 import org.apache.spark.sql.types.IntegerType$;
+import org.apache.spark.sql.types.Metadata;
```
Comment (d-m-h) [inline integration/spark/app/src/test/java/io/openlineage/spark/agent/lifecycle/plan/SaveIntoDataSourceCommandVisitorTest.java:39]:
```suggestion
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.Metadata;
```

**Review by d-m-h — COMMENTED** [review]
Mostly just getting rid of Scala-interop ugliness

File: integration/spark/app/src/test/java/io/openlineage/spark/agent/lifecycle/plan/SaveIntoDataSourceCommandVisitorTest.java (hunk 2 above)
Comment (d-m-h) [reply integration/spark/app/src/test/java/io/openlineage/spark/agent/lifecycle/plan/SaveIntoDataSourceCommandVisitorTest.java:68]: Now that I think about it, there are several ways to do this:

```java
    StructType schema = new StructType()
        .add("key", DataTypes.IntegerType, false, Metadata.empty())
        .add("value", DataTypes.StringType, false, Metadata.empty());
    
    List<StructField> structFields = Arrays.asList(
        DataTypes.createStructField("key", DataTypes.IntegerType, false),
        DataTypes.createStructField("value", DataTypes.StringType, false)
    );
    StructType schema = DataTypes.createStructType(structFields);
```

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/lifecycle/plan/SaveIntoDataSourceCommandVisitor.java (hunk 1 above)
Comment (ddebowczyk92) [reply integration/spark/shared/src/main/java/io/openlineage/spark/agent/lifecycle/plan/SaveIntoDataSourceCommandVisitor.java]: it's not, JDBCOptions class won't let it happen

File: integration/spark/app/src/test/java/io/openlineage/spark/agent/lifecycle/plan/SaveIntoDataSourceCommandVisitorTest.java (hunk 2 above)
Comment (ddebowczyk92) [reply integration/spark/app/src/test/java/io/openlineage/spark/agent/lifecycle/plan/SaveIntoDataSourceCommandVisitorTest.java]: applied version 1

**Review by mobuchowski — APPROVED** [review]

**Pull request merged**